File: scala-package/examples/src/main/scala/ml/dmlc/mxnetexamples/profiler/ProfilerNDArray.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ml.dmlc.mxnetexamples.profiler

import org.kohsuke.args4j.{CmdLineParser, Option}
import org.slf4j.LoggerFactory
import scala.collection.JavaConverters._
import java.io.File
import ml.dmlc.mxnet.Profiler
import ml.dmlc.mxnet.Random
import ml.dmlc.mxnet.Shape
import ml.dmlc.mxnet.NDArray
import ml.dmlc.mxnet.Context

/**
 * @author Depeng Liang
 */
object ProfilerNDArray {
  private val logger = LoggerFactory.getLogger(classOf[ProfilerNDArray])

  def testBroadcast(): Unit = {
    val sampleNum = 1000
    def testBroadcastTo(): Unit = {
      for (i <- 0 until sampleNum) {
        val nDim = scala.util.Random.nextInt(2) + 1
        val targetShape = Shape((0 until nDim).map(i => scala.util.Random.nextInt(10) + 1))
        val shape = targetShape.toArray.map { s =>
            if (scala.util.Random.nextInt(2) == 1) 1
            else s
        }
        val dat = NDArray.empty(shape: _*)
        val randomRet = (0 until shape.product)
          .map(r => scala.util.Random.nextFloat() - 0.5f).toArray
        dat.set(randomRet)
        val ndArrayRet = NDArray.broadcast_to(Map("shape" -> targetShape))(dat).get
        require(ndArrayRet.shape == targetShape)
        val err = {
          // implementation of broadcast
          val ret = {
            (randomRet /: shape.zipWithIndex.reverse){ (acc, elem) => elem match { case (s, i) =>
              if (s != targetShape(i)) {
                acc.grouped(shape.takeRight(shape.length - i).product).map {g =>
                  (0 until targetShape(i)).map(x => g).flatten
                }.flatten.toArray
              } else acc
            }}
          }
          val tmp = ndArrayRet.toArray.zip(ret).map{ case (l, r) => Math.pow(l - r, 2) }
          tmp.sum / tmp.length
        }
        require(err < 1E-8)
        ndArrayRet.dispose()
        dat.dispose()
      }
    }
    testBroadcastTo()
  }

  def randomNDArray(dim: Int): NDArray = {
    val tmp = Math.pow(1000, 1.0 / dim).toInt
    val shape = Shape((0 until dim).map(d => scala.util.Random.nextInt(tmp) + 1))
    Random.uniform(-10f, 10f, shape)
  }

  def testNDArraySaveload(): Unit = {
    val maxDim = 5
    val nRepeat = 10
    val fileName = s"${System.getProperty("java.io.tmpdir")}/tmpList.bin"
    for (repeat <- 0 until nRepeat) {
      try {
        val data = (0 until 10).map(i => randomNDArray(scala.util.Random.nextInt(4) + 1))
        NDArray.save(fileName, data)
        val data2 = NDArray.load2Array(fileName)
        require(data.length == data2.length)
        for ((x, y) <- data.zip(data2)) {
          val tmp = x - y
          require(tmp.toArray.sum == 0)
          tmp.dispose()
        }
        val dMap = data.zipWithIndex.map { case (arr, i) =>
          s"NDArray xx $i" -> arr
        }.toMap
        NDArray.save(fileName, dMap)
         val dMap2 = NDArray.load2Map(fileName)
         require(dMap.size == dMap2.size)
         for ((k, x) <- dMap) {
           val y = dMap2(k)
           val tmp = x - y
           require(tmp.toArray.sum == 0)
           tmp.dispose()
         }
        data.foreach(_.dispose())
      } finally {
        val file = new File(fileName)
        file.delete()
      }
    }
  }

  def testNDArrayCopy(): Unit = {
    val c = Random.uniform(-10f, 10f, Shape(10, 10))
    val d = c.copyTo(Context.cpu(0))
    val tmp = c - d
    require(tmp.toArray.map(Math.abs).sum == 0)
    c.dispose()
    d.dispose()
  }

  def reldiff(a: NDArray, b: NDArray): Float = {
    val diff = NDArray.sum(NDArray.abs(a - b)).toScalar
    val norm = NDArray.sum(NDArray.abs(a)).toScalar
    diff / norm
  }

  def reldiff(a: Array[Float], b: Array[Float]): Float = {
    val diff =
      (a zip b).map { case (aElem, bElem) => Math.abs(aElem - bElem) }.sum
    val norm: Float = a.reduce(Math.abs(_) + Math.abs(_))
    diff / norm
  }

  def testNDArrayNegate(): Unit = {
    val rand = Random.uniform(-10f, 10f, Shape(2, 3, 4))
    val npy = rand.toArray
    val arr = NDArray.empty(Shape(2, 3, 4))
    arr.set(npy)
    require(reldiff(npy, arr.toArray) < 1e-6f)
    val negativeArr = -arr
    require(reldiff(npy.map(_ * -1f), negativeArr.toArray) < 1e-6f)
    // a final check to make sure the negation (-) is not implemented
    // as inplace operation, so the contents of arr does not change after
    // we compute (-arr)
    require(reldiff(npy, arr.toArray) < 1e-6f)
    rand.dispose()
    arr.dispose()
    negativeArr.dispose()
  }

  def testNDArrayScalar(): Unit = {
    val c = NDArray.empty(10, 10)
    val d = NDArray.empty(10, 10)
    c.set(0.5f)
    d.set(1.0f)
    d -= c * 2f / 3f * 6f
    c += 0.5f
    require(c.toArray.sum - 100f < 1e-5f)
    require(d.toArray.sum + 100f < 1e-5f)
    c.set(2f)
    require(c.toArray.sum - 200f < 1e-5f)
    d.set(-c + 2f)
    require(d.toArray.sum < 1e-5f)
    c.dispose()
    d.dispose()
  }

  def testClip(): Unit = {
    val shape = Shape(10)
    val A = Random.uniform(-10f, 10f, shape)
    val B = NDArray.clip(A, -2f, 2f)
    val B1 = B.toArray
    require(B1.forall { x => x >= -2f && x <= 2f })
  }

  def testDot(): Unit = {
    val a = Random.uniform(-3f, 3f, Shape(3, 4))
    val b = Random.uniform(-3f, 3f, Shape(4, 5))
    val c = NDArray.dot(a, b)
    val A = a.toArray.grouped(4).toArray
    val B = b.toArray.grouped(5).toArray
    val C = (Array[Array[Float]]() /: A)((acc, row) => acc :+ row.zip(B).map(z =>
                z._2.map(_ * z._1)).reduceLeft(_.zip(_).map(x => x._1 + x._2))).flatten
    require(reldiff(c.toArray, C) < 1e-5f)
    a.dispose()
    b.dispose()
    c.dispose()
  }

  def testNDArrayOnehot(): Unit = {
    val shape = Shape(100, 20)
    var npy = (0 until shape.product).toArray.map(_.toFloat)
    val arr = NDArray.empty(shape)
    arr.set(npy)
    val nRepeat = 3
    for (repeat <- 0 until nRepeat) {
      val indices = (0 until shape(0)).map(i => scala.util.Random.nextInt(shape(1)))
      npy = npy.map(i => 0f)
      for (i <- 0 until indices.length) npy(i * shape(1) + indices(i)) = 1f
      val ind = NDArray.empty(shape(0))
      ind.set(indices.toArray.map(_.toFloat))
      NDArray.onehotEncode(ind, arr)
      require(arr.toArray.zip(npy).map(x => x._1 - x._2).sum == 0f)
      ind.dispose()
    }
    arr.dispose()
  }

  def main(args: Array[String]): Unit = {
    val eray = new ProfilerNDArray
    val parser: CmdLineParser = new CmdLineParser(eray)
    try {
      parser.parseArgument(args.toList.asJava)

      val path = s"${eray.outputPath}${File.separator}${eray.profilerName}"
      Profiler.profilerSetConfig(mode = eray.profilerMode, fileName = path)
      logger.info(s"profile file save to $path")

      Profiler.profilerSetState("run")
      testBroadcast()
      testNDArraySaveload()
      testNDArrayCopy()
      testNDArrayNegate()
      testNDArrayScalar()
      testClip()
      testDot()
      testNDArrayOnehot()
      Profiler.profilerSetState("stop")

    } catch {
      case ex: Exception => {
        logger.error(ex.getMessage, ex)
        parser.printUsage(System.err)
        sys.exit(1)
      }
    }
  }
}

class ProfilerNDArray {
  @Option(name = "--profiler-mode", usage = "the profiler mode, can be \"symbolic\" or \"all\".")
  private val profilerMode: String = "all"
  @Option(name = "--output-path", usage = "the profile file output directory.")
  private val outputPath: String = "."
  @Option(name = "--profile-filename", usage = "the profile file name.")
  private val profilerName: String = "profile_ndarray.json"
}

File: example/profiler/profiler_ndarray.py
import os
import mxnet as mx
import numpy as np
import pickle as pkl


def _np_reduce(dat, axis, keepdims, numpy_reduce_func):
    if isinstance(axis, int):
        axis = [axis]
    else:
        axis = list(axis) if axis is not None else range(len(dat.shape))
    ret = dat
    for i in reversed(sorted(axis)):
        ret = numpy_reduce_func(ret, axis=i)
    if keepdims:
        keepdims_shape = list(dat.shape)
        for i in axis:
            keepdims_shape[i] = 1
        ret = ret.reshape(tuple(keepdims_shape))
    return ret


def reldiff(a, b):
    diff = np.abs(a - b)
    norm = np.abs(a)
    reldiff = np.max(diff  / (norm + 1e-7))
    return reldiff


def same(a, b):
    return np.sum(a != b) == 0


def check_with_uniform(uf, arg_shapes, dim=None, npuf=None, rmin=-10, type_list=[np.float32]):
    """check function consistency with uniform random numbers"""
    if isinstance(arg_shapes, int):
        assert dim
        shape = tuple(np.random.randint(1, int(1000**(1.0/dim)), size=dim))
        arg_shapes = [shape] * arg_shapes
    for dtype in type_list:
        ndarray_arg = []
        numpy_arg = []
        for s in arg_shapes:
            npy = np.random.uniform(rmin, 10, s).astype(dtype)
            narr = mx.nd.array(npy, dtype=dtype)
            ndarray_arg.append(narr)
            numpy_arg.append(npy)
        out1 = uf(*ndarray_arg)
        if npuf is None:
            out2 = uf(*numpy_arg).astype(dtype)
        else:
            out2 = npuf(*numpy_arg).astype(dtype)

        assert out1.shape == out2.shape
        if isinstance(out1, mx.nd.NDArray):
            out1 = out1.asnumpy()
        if dtype == np.float16:
            assert reldiff(out1, out2) < 2e-3
        else:
            assert reldiff(out1, out2) < 1e-6


def random_ndarray(dim):
    shape = tuple(np.random.randint(1, int(1000**(1.0/dim)), size=dim))
    data = mx.nd.array(np.random.uniform(-10, 10, shape))
    return data

def test_ndarray_elementwise():
    np.random.seed(0)
    nrepeat = 10
    maxdim = 4
    all_type = [np.float32, np.float64, np.float16, np.uint8, np.int32]
    real_type = [np.float32, np.float64, np.float16]
    for repeat in range(nrepeat):
        for dim in range(1, maxdim):
            check_with_uniform(lambda x, y: x + y, 2, dim, type_list=all_type)
            check_with_uniform(lambda x, y: x - y, 2, dim, type_list=all_type)
            check_with_uniform(lambda x, y: x * y, 2, dim, type_list=all_type)
            check_with_uniform(lambda x, y: x / y, 2, dim, type_list=real_type)
            check_with_uniform(lambda x, y: x / y, 2, dim, rmin=1, type_list=all_type)
            check_with_uniform(mx.nd.sqrt, 1, dim, np.sqrt, rmin=0)
            check_with_uniform(mx.nd.square, 1, dim, np.square, rmin=0)
            check_with_uniform(lambda x: mx.nd.norm(x).asscalar(), 1, dim, np.linalg.norm)

def test_ndarray_negate():
    npy = np.random.uniform(-10, 10, (2,3,4))
    arr = mx.nd.array(npy)
    assert reldiff(npy, arr.asnumpy()) < 1e-6
    assert reldiff(-npy, (-arr).asnumpy()) < 1e-6

    # a final check to make sure the negation (-) is not implemented
    # as inplace operation, so the contents of arr does not change after
    # we compute (-arr)
    assert reldiff(npy, arr.asnumpy()) < 1e-6


def test_ndarray_choose():
    shape = (100, 20)
    npy = np.arange(np.prod(shape)).reshape(shape)
    arr = mx.nd.array(npy)
    nrepeat = 3
    for repeat in range(nrepeat):
        indices = np.random.randint(shape[1], size=shape[0])
        assert same(npy[np.arange(shape[0]), indices],
                    mx.nd.choose_element_0index(arr, mx.nd.array(indices)).asnumpy())


def test_ndarray_fill():
    shape = (100, 20)
    npy = np.arange(np.prod(shape)).reshape(shape)
    arr = mx.nd.array(npy)
    new_npy = npy.copy()
    nrepeat = 3
    for repeat in range(nrepeat):
        indices = np.random.randint(shape[1], size=shape[0])
        val = np.random.randint(shape[1], size=shape[0])
        new_npy[:] = npy
        new_npy[np.arange(shape[0]), indices] = val
        assert same(new_npy,
                    mx.nd.fill_element_0index(arr, mx.nd.array(val), mx.nd.array(indices)).asnumpy())


def test_ndarray_onehot():
    shape = (100, 20)
    npy = np.arange(np.prod(shape)).reshape(shape)
    arr = mx.nd.array(npy)
    nrepeat = 3
    for repeat in range(nrepeat):
        indices = np.random.randint(shape[1], size=shape[0])
        npy[:] = 0.0
        npy[np.arange(shape[0]), indices] = 1.0
        mx.nd.onehot_encode(mx.nd.array(indices), out=arr)
        assert same(npy, arr.asnumpy())


def test_ndarray_copy():
    c = mx.nd.array(np.random.uniform(-10, 10, (10, 10)))
    d = c.copyto(mx.Context('cpu', 0))
    assert np.sum(np.abs(c.asnumpy() != d.asnumpy())) == 0.0


def test_ndarray_scalar():
    c = mx.nd.empty((10,10))
    d = mx.nd.empty((10,10))
    c[:] = 0.5
    d[:] = 1.0
    d -= c * 2 / 3 * 6.0
    c += 0.5
    assert(np.sum(c.asnumpy()) - 100 < 1e-5)
    assert(np.sum(d.asnumpy()) + 100 < 1e-5)
    c[:] = 2
    assert(np.sum(c.asnumpy()) - 200 < 1e-5)
    d = -c + 2
    assert(np.sum(d.asnumpy()) < 1e-5)

def test_ndarray_pickle():
    np.random.seed(0)
    maxdim = 5
    nrepeat = 10
    for repeat in range(nrepeat):
        for dim in range(1, maxdim):
            a = random_ndarray(dim)
            b = mx.nd.empty(a.shape)
            a[:] = np.random.uniform(-10, 10, a.shape)
            b[:] = np.random.uniform(-10, 10, a.shape)
            a = a + b
            data = pkl.dumps(a)
            a2 = pkl.loads(data)
            assert np.sum(a.asnumpy() != a2.asnumpy()) == 0


def test_ndarray_saveload():
    np.random.seed(0)
    maxdim = 5
    nrepeat = 10
    fname = 'tmp_list.bin'
    for repeat in range(nrepeat):
        data = []
        for i in range(10):
            data.append(random_ndarray(np.random.randint(1, 5)))
        mx.nd.save(fname, data)
        data2 = mx.nd.load(fname)
        assert len(data) == len(data2)
        for x, y in zip(data, data2):
            assert np.sum(x.asnumpy() != y.asnumpy()) == 0
        dmap = {'ndarray xx %s' % i : x for i, x in enumerate(data)}
        mx.nd.save(fname, dmap)
        dmap2 = mx.nd.load(fname)
        assert len(dmap2) == len(dmap)
        for k, x in dmap.items():
            y = dmap2[k]
            assert np.sum(x.asnumpy() != y.asnumpy()) == 0
    os.remove(fname)


def test_ndarray_slice():
    shape = (10,)
    A = mx.nd.array(np.random.uniform(-10, 10, shape))
    A2 = A.asnumpy()
    assert same(A[3:8].asnumpy(), A2[3:8])
    A2[3:8] *= 10;
    A[3:8] = A2[3:8]
    assert same(A[3:8].asnumpy(), A2[3:8])


def test_ndarray_slice_along_axis():
    arr = mx.nd.array(np.random.uniform(-10, 10, (3, 4, 2, 3)))
    sub_arr = mx.nd.zeros((3, 2, 2, 3))
    arr._copy_slice_to(1, 1, 3, sub_arr)

    # test we sliced correctly
    assert same(arr.asnumpy()[:, 1:3, :, :], sub_arr.asnumpy())

    # test that slice is copy, instead of shared memory
    sub_arr[:] = 0
    assert not same(arr.asnumpy()[:, 1:3, :, :], sub_arr.asnumpy())


def test_clip():
    shape = (10,)
    A = mx.random.uniform(-10, 10, shape)
    B = mx.nd.clip(A, -2, 2)
    B1 = B.asnumpy()
    for i in range(shape[0]):
        assert B1[i] >= -2
        assert B1[i] <= 2

def test_dot():
    a = np.random.uniform(-3, 3, (3, 4))
    b = np.random.uniform(-3, 3, (4, 5))
    c = np.dot(a, b)
    A = mx.nd.array(a)
    B = mx.nd.array(b)
    C = mx.nd.dot(A, B)
    assert reldiff(c, C.asnumpy()) < 1e-5

def test_reduce():
    sample_num = 200
    def test_reduce_inner(numpy_reduce_func, nd_reduce_func):
        for i in range(sample_num):
            ndim = np.random.randint(1, 6)
            shape = np.random.randint(1, 11, size=ndim)
            axis_flags = np.random.randint(0, 2, size=ndim)
            axes = []
            for (axis, flag) in enumerate(axis_flags):
                if flag:
                    axes.append(axis)
            keepdims = np.random.randint(0, 2)
            dat = np.random.rand(*shape) - 0.5
            if 0 == len(axes):
                axes = tuple(range(ndim))
            else:
                axes = tuple(axes)
            numpy_ret = numpy_reduce_func(dat, axis=axes, keepdims=keepdims)

            ndarray_ret = nd_reduce_func(mx.nd.array(dat), axis=axes, keepdims=keepdims)
            if type(ndarray_ret) is mx.ndarray.NDArray:
                ndarray_ret = ndarray_ret.asnumpy()
            assert (ndarray_ret.shape == numpy_ret.shape) or \
                   (ndarray_ret.shape == (1,) and numpy_ret.shape == ()), "nd:%s, numpy:%s" \
                                                         %(ndarray_ret.shape, numpy_ret.shape)
            err = np.square(ndarray_ret - numpy_ret).mean()
            assert err < 1E-4
    test_reduce_inner(lambda data, axis, keepdims:_np_reduce(data, axis, keepdims, np.sum),
                      mx.nd.sum)
    test_reduce_inner(lambda data, axis, keepdims:_np_reduce(data, axis, keepdims, np.max),
                      mx.nd.max)
    test_reduce_inner(lambda data, axis, keepdims:_np_reduce(data, axis, keepdims, np.min),
                      mx.nd.min)

def test_broadcast():
    sample_num = 1000
    def test_broadcast_to():
        for i in range(sample_num):
            ndim = np.random.randint(1, 6)
            target_shape = np.random.randint(1, 11, size=ndim)
            shape = target_shape.copy()
            axis_flags = np.random.randint(0, 2, size=ndim)
            axes = []
            for (axis, flag) in enumerate(axis_flags):
                if flag:
                    shape[axis] = 1
            dat = np.random.rand(*shape) - 0.5
            numpy_ret = dat
            ndarray_ret = mx.nd.array(dat).broadcast_to(shape=target_shape)
            if type(ndarray_ret) is mx.ndarray.NDArray:
                ndarray_ret = ndarray_ret.asnumpy()
            assert (ndarray_ret.shape == target_shape).all()
            err = np.square(ndarray_ret - numpy_ret).mean()
            assert err < 1E-8
    test_broadcast_to()

if __name__ == '__main__':
    mx.profiler.profiler_set_config(mode='all', filename='profile_ndarray.json')
    mx.profiler.profiler_set_state('run')
    test_ndarray_slice_along_axis()
    test_broadcast()
    test_ndarray_elementwise()
    test_ndarray_slice()
    test_ndarray_pickle()
    test_ndarray_saveload()
    test_ndarray_copy()
    test_ndarray_negate()
    test_ndarray_scalar()
    test_clip()
    test_dot()
    test_ndarray_choose()
    test_ndarray_onehot()
    test_ndarray_fill()
    test_reduce()
    mx.profiler.profiler_set_state('stop')
